numpy/random/mtrand/mtrand.pyx

```diff
@@ -959,31 +961,77 @@ cdef class RandomState:
         array([[4, 0, 2, 1],
                [3, 2, 2, 0]])
 
+        Generate a 1 x 3 array with 3 different upper bounds:
+        >>> np.random.randint(1, [3, 5, 10])
+        array([2, 2, 9])
+
+        Generate a 1 x 3 array with 3 different lower bounds:
+        >>> np.random.randint([1, 5, 7], 10)
+        array([9, 8, 7])
+
+        Generate a 1 x 3 array with 3 different lower and upper bounds:
+        >>> np.random.randint([1, 5, 7], [13, 30, 45])
+        array([12,  5, 18])
+
+        Generate a 2 x 4 array of ints of the data type np.uint8:
+        >>>> np.random.randint(np.iinfo(np.uint8).min,
+                           np.iinfo(np.uint8).max + 1,
+                           size=(2, 4), dtype=np.uint8)
+        array([[ 35, 252,  90, 214],
+               [ 47,  23, 214, 185]], dtype=uint8)
+
         """
         if high is None:
             high = low
             low = 0
 
+        # '_randint_type' is defined in
+        # 'generate_randint_helpers.py'
         key = np.dtype(dtype).name
         if not key in _randint_type:
             raise TypeError('Unsupported dtype "%s" for randint' % key)
-        lowbnd, highbnd, randfunc = _randint_type[key]
 
-        if low < lowbnd:
-            raise ValueError("low is out of bounds for %s" % (key,))
-        if high > highbnd:
-            raise ValueError("high is out of bounds for %s" % (key,))
-        if low >= high:
-            raise ValueError("low >= high")
+        lowbnd, highbnd, func = _randint_type[key]
 
-        with self.lock:
-            ret = randfunc(low, high - 1, size, self.state_address)
+        low = np.array(low)
+        high = np.array(high)
 
-            if size is None:
-                if dtype in (np.bool, np.int, np.long):
-                    return dtype(ret)
+        if low.shape == high.shape == ():
+            ilow = int(low)
+            ihigh = int(high)
 
-            return ret
+            if ilow < lowbnd:
+                raise ValueError("low is out of bounds for %s" % (key,))
+            if ihigh > highbnd:
+                raise ValueError("high is out of bounds for %s" % (key,))
+            if ilow >= ihigh:
+                raise ValueError("low >= high")
+
+            with self.lock:
+                ret = func(ilow, ihigh - 1, size, self.state_address)
+
+                if size is None:
+                    if dtype in (np.bool, np.int, np.long):
+                        return dtype(ret)
+
+                return ret
+        else:
+            if np.any(np.less(low, lowbnd)):
+                raise ValueError("low is out of bounds for %s" % (key,))
+            if np.any(np.greater(high, highbnd)):
+                raise ValueError("high is out of bounds for %s" % (key,))
+
+            # We have to check whether `low` >= `high` in both ways
+            # to avoid issues of overflow and rounding when comparing
+            # an np.uint64 against np.iinfo(np.int64).max
```

Comment:
I think the justification is rather straightforward: `randint` generates numbers between `[a, b)`.  When both are provided, we check that `[a, b - 1]` fits the `dtype` specified.  Otherwise, we error.  This is how it has been designed from the beginning.  However, it was not robust to extreme points in the ranges of `int64`, which is what these tests are designed for.  Thus, they are in fact very consistent with the API here.

What part of this do you not find desirable, and more importantly, why?